File: python/ch07/03.subgraph-function.py
from typing import TypedDict
from langgraph.graph import START, StateGraph

class State(TypedDict):
    foo: str

class SubgraphState(TypedDict):
    # 부모 그래프와 키를 공유하지 않음
    bar: str
    baz: str


# 서브그래프 정의
def subgraph_node(state: SubgraphState):
    return {"bar": state["bar"] + "baz"}


subgraph_builder = StateGraph(SubgraphState)
subgraph_builder.add_node("subgraph_node", subgraph_node)
subgraph_builder.add_edge(START, "subgraph_node")
# 서브그래프에 필요한 추가 설정은 여기에 작성
subgraph = subgraph_builder.compile()


# 서브그래프를 호출하는 부모 그래프 정의
def node(state: State):
    # 부모 그래프의 상태를 서브그래프 상태로 변환
    response = subgraph.invoke({"bar": state["foo"]})
    # 응답을 다시 부모 그래프의 상태로 변환
    return {"foo": response["bar"]}


builder = StateGraph(State)
# 서브그래프 대신 `node`를 지정
builder.add_node("node", node)
builder.add_edge(START, "node")
# 부모 그래프에 필요한 추가 설정은 여기에 작성
graph = builder.compile()

# 예시
initial_state = {"foo": "hello"}
result = graph.invoke(initial_state)
print(
    f"Result: {result}"
)  # foo를 bar로 변환해 "baz"를 추가하고 다시 foo로 변환

File: javascript/ch07/03.subgraph-function.js
import { StateGraph, START, Annotation } from '@langchain/langgraph';

const StateAnnotation = Annotation.Root({
  foo: Annotation(),
});

const SubgraphStateAnnotation = Annotation.Root({
  // 부모 그래프와 키를 공유하지 않음
  bar: Annotation(),
  baz: Annotation(),
});

// 서브그래프 정의
const subgraphNode = async (state) => {
  return { bar: state.bar + 'baz' };
};

const subgraph = new StateGraph(SubgraphStateAnnotation)
  .addNode('subgraph', subgraphNode)
  .addEdge(START, 'subgraph')
  // 서브그래프에 필요한 추가 설정은 여기에 작성
  .compile();

// 서브그래프를 호출하는 부모 그래프 정의
const subgraphWrapperNode = async (state) => {
  // 부모 그래프의 상태를 서브그래프 상태로 변환
  const response = await subgraph.invoke({
    bar: state.foo,
  });
  // 응답을 다시 부모 그래프의 상태로 변환
  return {
    foo: response.bar,
  };
};

const parentGraph = new StateGraph(StateAnnotation)
  // `subgraph` 대신 `subgraphWrapperNode`를 지정
  .addNode('subgraph', subgraphWrapperNode)
  .addEdge(START, 'subgraph')
  // 부모 그래프에 필요한 추가 설정은 여기에 작성
  .compile();

// 예시

const initialState = { foo: 'hello' };
const result = await parentGraph.invoke(initialState);
console.log(`Result: ${JSON.stringify(result)}`); // foo를 bar로 변환해 "baz"를 추가하고 다시 foo로 변환
